### ISSUE TITLE
Question about theta_class (\theta^{\rm mixed} in you paper)

### ISSUE THREAD

**Issue by ftft33** [wilkinghoff/sub-cluster-AdaCos#3]
@wilkinghoff 

Thanks for sharing your code for Sub-Cluster Adacos.
Would it be possible for you to add an explanation of \theta^{\rm mixed} (eq.(8) in your paper)?

My understanding is that \theta^{\rm mixed} is implemented as the sum of the mixed angles for all subclusters multiplied by count_nonzero. Is this correct?

I couldn't understand why you implemented Eq.(8) in this way, and I would be grateful if you could explain this processing.

`theta_class = tf.reduce_sum(y1 * theta, axis=1) * tf.math.count_nonzero(y1_orig, axis=1, dtype=tf.dtypes.float32)  # take mix-upped angle of mix-upped classes`

**Comment by wilkinghoff** [comment]
Thank you for your interest and pointing this out. You are right, it seems to be an error and probably is an artifact caused by changing the code back and forth during the experimentation phase. I must admit that I don't have another explanation.

The dimension of y1 and theta is (n_samples, n_classes*n_subclusters) and y1 corresponds to lambda in Eq. (8).  In Eq. (8) the angle between a sample and its class is the sum of all angles between the sample and the centers of the sub-clusters. So, `tf.reduce_sum(y1 * theta, axis=1)` already results in a mixed-up angle for each sample and according to Eq. (8) the second part, i.e. the scaling with `tf.math.count_nonzero(y1_orig, axis=1, dtype=tf.dtypes.float32)`, should be removed.

Sorry for the inconvenience. I will take a closer look as soon as I find the time to do so.

**Comment by ftft33** [comment]
@wilkinghoff 
Thank you for your quick reply!!!
I understood it from your answer, but I have another question.
Is it correct to calculate the sum of all angles for all sub-clusters as `\theta^{\rm mixed}`?
Is my understanding correct that theta_class (`tf.reduce_sum(y1 * theta, axis=1)`) would exceed [0, pi] if the number of sub-clusters increases and `\frac{\pi}{4}` would be selected in `\min(\frac{\pi}{4}, \theta_{\rm med}`)?

**Comment by wilkinghoff** [comment]
Thanks again for your interest and having this discussion. This is a good point.

> Is it correct to calculate the sum of all angles for all sub-clusters as \theta^{\rm mixed}?

It is hard to say if it is 'correct' or not because it is just a definition without any proven properties. In the experiments for this paper, this led to slightly better ASD results. But there may very well be a better choice for setting the scale parameter.

> Is my understanding correct that theta_class (tf.reduce_sum(y1 * theta, axis=1)) would exceed [0, pi] if the number of sub-clusters increases and \frac{\pi}{4} would be selected in \min(\frac{\pi}{4}, \theta_{\rm med})?

You are absolutely right. Selecting `\frac{\pi}{4}` in these cases is a good thing to still ensure sensitivity to the angles. The intuitive idea of introducing the sub-clusters was that these resemble overlapping radial basis functions to allow the network to learn more sophisticated distributions than just a single Gaussian, which corresponds to a single center. If the angles to all sub-clusters are relatively small, which is possible if the sub-clusters are close to each other after training for a bit, `\theta_{\rm med}` would still be selected. Again, this choice may very well not be optimal. Still, it seemed to work in the context of the paper.

I hope that I was able to answer your questions.

**Comment by ftft33** [comment]
@wilkinghoff 
Thank you so much for your clear explanation!
I was able to deepen my understanding of your work.

> It is hard to say if it is 'correct' or not because it is just a definition without any proven properties.

I was aware of that aspect, but I wanted to know whether this implementation was intended by you. Anyway, all of my questions are resolved by your answer.
Many thanks once again!!!
